File: client/src/main/java/dev/hoot/api/entities/Projectiles.java
package dev.hoot.api.entities;

import dev.hoot.api.commons.Predicates;
import dev.hoot.api.game.Game;
import meteor.Main;
import net.runelite.api.Actor;
import net.runelite.api.Projectile;
import net.runelite.api.coords.LocalPoint;
import net.runelite.api.coords.WorldPoint;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;
import java.util.function.Predicate;

public class Projectiles
{
	private Projectiles()
	{
	}

	public static List<Projectile> getAll(Predicate<Projectile> filter)
	{
		List<Projectile> out = new ArrayList<>();
		Game.getClient().getProjectiles().forEach(p ->
		{
			if (filter.test(p))
			{
				out.add(p);
			}
		});

		return out;
	}

	public static List<Projectile> getAll(int... ids)
	{
		return getAll(Predicates.ids(ids));
	}

	public static Projectile getNearest(Predicate<Projectile> filter)
	{
		return getAll(filter).stream()
				.min(Comparator.comparingInt(p ->
						WorldPoint.fromLocal(Game.getClient(), (int) p.getX(), (int) p.getY(), Game.getClient().getPlane())
								.distanceTo(Main.client.getLocalPlayer().getWorldLocation()))
				)
				.orElse(null);
	}

	public static Projectile getNearest(int... ids)
	{
		return getNearest(Predicates.ids(ids));
	}

	public static Projectile getNearest(Actor target)
	{
		return getNearest(x -> x.getInteracting() != null && x.getInteracting().equals(target));
	}

	public static Projectile getNearest(WorldPoint startPoint)
	{
		LocalPoint localPoint = LocalPoint.fromWorld(Game.getClient(), startPoint);
		if (localPoint == null)
		{
			return null;
		}

		return getNearest(x -> x.getX1() == localPoint.getX() && x.getY1() == localPoint.getY());
	}
}

File: src/main/kotlin/meteor/api/entities/Projectiles.kt
package meteor.api.entities

import Main.client

import net.runelite.api.Actor
import net.runelite.api.Projectile
import net.runelite.api.coords.LocalPoint
import net.runelite.api.coords.WorldPoint
import java.util.Comparator
import java.util.function.Predicate
import java.util.function.ToIntFunction
import java.util.stream.Collectors

object Projectiles {
    fun getAll(filter: Predicate<Projectile>): List<Projectile?> {
        return client.getProjectiles().stream()
            .filter { projectile -> projectile != null && filter.test(projectile) }
            .collect(Collectors.toList())
    }

    fun getAll(vararg ids: Int): List<Projectile?> {
        return getAll { x: Projectile ->
            for (id in ids) {
                if (id == x.id) {
                    return@getAll true
                }
            }
            false
        }
    }

    fun getNearest(filter: Predicate<Projectile>): Projectile? {
        return getAll(filter).stream()
            .min(
                Comparator.comparingInt { p: Projectile? ->
                    WorldPoint.fromLocal(client, p!!.x.toInt(), p.y.toInt(), client.getPlane())
                        .distanceTo(Players.local.worldLocation)
                }
            )
            .orElse(null)
    }

    fun getNearest(vararg ids: Int): Projectile? {
        return getNearest { x: Projectile ->
            for (id in ids) {
                if (id == x.id) {
                    return@getNearest true
                }
            }
            false
        }
    }

    fun getNearest(target: Actor): Projectile? {
        return getNearest { x: Projectile -> x.interacting != null && x.interacting == target }
    }

    fun getNearest(startPoint: WorldPoint?): Projectile? {
        val localPoint = LocalPoint.fromWorld(client, startPoint) ?: return null
        return getNearest { x: Projectile -> x.x1 == localPoint.x && x.y1 == localPoint.y }
    }
}